### ISSUE TITLE
Helper ignoring coming action when previous action is still running

### ISSUE THREAD

**Issue by tomoyukim** [redux-saga/redux-saga#1315]
Hi, I locally created a helper named `takeOnce` like below in order to ignore coming action when the previous action is still running.

```javascript
function* takeOnce(pattern, saga, ...args) {
  const lastTask;
  while (true) {
    const action = yield take(pattern);
    if (_.isNil(lastTask) || !lastTask.isRunning()) {
      lastTask = yield fork(saga, ...args.concat(action));
    }
  }
}
```

I'm happy if it's supported by the library itself as authorized helper and if this is helpful, I'd like to create a new PR. So, what do you think about it?

**Comment by Andarist** [comment]
`takeOnce` would mean for me "take once and exit", so I'd appreciate a better (more descriptive) name. It seems like you have a kinda `leadingDebounce` mechanism which takes only a leading action in a sequence until it finishes a task - `debounce` most often is parametrized with some time parameter, so it's probably not a best name for this though.

It can also be implemented a little bit easier (without introspecting `task`):
```js
function* GIVE_ME_NAME(pattern, saga, ...args) {
	cosnt chan = actionChannel(pattern, buffers.dropping(1))
	while (true) {
		const action = yield take(chan)
		yield call(saga, ...args, action)
		yield flush(chan)
	}
}
```

If we can come up with a better name I would be glad to merge this addition. Keep in mind though that internally helpers (takeEvery, takeLatest, throttle) are not implemented with generators syntax, you'd have to implement this with internal `fsmIterator` - if you need any help with understanding how it works I can explain it.

**Comment by tomoyukim** [comment]
Thank you for your positive and kind feedback.

> If we can come up with a better name I would be glad to merge this addition.

Another idea comes to mind is `takeExclusive` or `takeExclusively` because I think it looks like somehow mutual exclusion. What about this?

I've tried to write code with `fsmIterator` referring another `take*` helpers based on my above snippet but I feel your implementation is better than me. So, I'll try to rewrite it with your idea.

BTW, I guess I need to include testing and maybe document when I create PR. It seems contribution guid is under discussion according to https://github.com/redux-saga/redux-saga/issues/1287. So, would you let me know what I should do for the PR?

**Comment by Andarist** [comment]
Let's start with providing a `fsmIterator`-based implementation and tests. Docs would ofc be great too - adding it to the API docs with an example usage would be superb.

**Comment by tomoyukim** [comment]
OK. Thank you for your support. I'll prepare for the PR.

**Comment by Andarist** [comment]
Actually I have posted even simpler (I think its the same use case) implementation for this once - https://github.com/redux-saga/redux-saga/issues/589#issuecomment-348344469

**Comment by tomoyukim** [comment]
All right. Thank you for your update. I'll improve my local branch with it.

**Comment by tomoyukim** [comment]
@Andarist I've created PR for `takeLeading` as following another discussion. Thank you very much for your great support and waiting.

**Comment by atlanteh** [comment]
I don't see it in latest release. I'm assuming it hasn't been published yet to npm? Anyone can publish a new version?

**Comment by Andarist** [comment]
@atlanteh It will be published soon-ish as v1.0.0-beta.1
